chapters/lexicalstructure.tex

```diff
@@ -247,6 +247,24 @@ \subsection{Strings}\label{strings}
 \end{lstlisting}
 \end{nonnormative}
 
+\subsection{Units of Literal Constants}\label{units-literal-constants}
+
+Literal \lstinline!Real! and \lstinline!Integer! constants are implicitly assumed to have \lstinline!unit = "1"! if they appear in expressions containing variables, parameters, or constants that have a non-empty \lstinline!unit! attribute string.
```

Comment:
I am skeptical about the vagueness of _if they appear in expressions containing variables_.  For example, the way I read it would cause a unit error in the function call here:
```
model FunCall
  function f
    input Real u(unit = "m");
    output Real y(unit = "m") = u;
  end f;

  Real x(unit = "m") = 1.0;
  Real y(unit = "m") = x + f(2.0);
end FunCall;
```
Here, `x + f(1.0)` is an expression involving a variable with non-empty `unit`, and hence `2.0` would get the unit `"1"` which is not compatible with `f`.

Reply:
That's a good point. It was fairly obvious to me that function calls act like "shields", and what matters is the unit of the function's output, not the unit of the function input expression; on the other hand, one should handle Real literals in expressions passed as function inputs in the same way. But I agree that normative text should be precise and not require  interpretation, which is ambiguous by definition.

We could write something like this:

> Literal `Real` and `Integer` constants are implicitly assumed to have `unit = "1"` if they appear in expressions containing variables, parameters, and constants that have a non-empty `unit` attribute string, or function calls that have a non-empty `unit` attribute of the output variable. The same rule applies to function call inputs.

In fact, if we want to be consistent with Hans's previous advice on transcendental functions inputs and outputs, we could write
> Literal `Real` and `Integer` constants are implicitly assumed to have _a dimensionless unit_ if they appear in expressions containing variables, parameters, and constants that have a non-empty `unit` attribute string, or function calls that have a non-empty `unit` attribute of the output variable. The same rule applies to function call inputs.

Shall I commit the second variant?

Reply:
See below for a commit-ready suggestion

Reply:
> That's a good point. It was fairly obvious to me that function calls act like "shields", and what matters is the unit of the function's output, not the unit of the function input expression; on the other hand, one should handle Real literals in expressions passed as function inputs in the same way. 

I think this can be resolved by thinking about the parse tree of expressions. I think what Francesco meant with "appear"
> if they appear in expressions containing...

is *if they are an immediate child of expressions containing...*

When walking up the tree of expression, an expression only "sees" its immediate children. 
In the case of the "shielding" function, the function call of `f` is a separate node in the tree: Its only child is its argument, so we need to define how a function call reconciles a real/int argument with a unitful parameter.
Then, for the purposes of the expression `x + f(2.0);`, its children are `x` (unit "m"), and the _output_ of `f` (unit "m"), so everything works out. :-)

Reply:
The `2.0` in `f(2.0)` gets unit `""`, so the function call is OK, as intended. 
However, in the case of 
```modelica
model FunCall2
  function g
    input Real u1(unit = "m");
    input Real u2(unit = "m");
    output Real y(unit = "m") = u1 + u2;
  end g;

  Real x(unit = "m") = 1.0;
  Real y(unit = "m") = f(2.0, x);
end FunCall;
```
we run into a problem in `f(2.0, x)`: As `x` is a *variable that has a non-empty unit attribute string*, in this expression `2.0` is a literal Real constant that appears in an expression with a variable that has a non-empty unit attribute string, so `2.0` will get a unit of "1", which clashes with the unit of `g`'s input argument `u1` ("m").

So, "appear" seems still not precise enough. Right now I don't have a better proposal, maybe someone else?

Reply:
I think with the new approach to formulation, this conversation can be resolved, because we don't use the "appear" formulation anymore.